### ISSUE TITLE
check that diamond storage is untouched after each pr

### ISSUE THREAD

**Issue by zgorizzo69** [pavlovcik/uad-ui#636]
best would be to add a gh action to make sure that we haven't messed with the diamond storage

**Comment by PhantomCracker** [comment]
Hello @zgorizzo69 ,
Can you help with something? I have 2 questions:
1. where can i find the diamond storage?
2. is there a backup for diamond storage? 

I am thinking of checking the diamond storage before creating a PR and after

**Comment by PhantomCracker** [comment]
This is work in progress

**Comment by 0x4007** [comment]
> Hello @zgorizzo69 , Can you help with something? I have 2 questions:
> 
> 1. where can i find the diamond storage?
> 2. is there a backup for diamond storage?
> 
> I am thinking of checking the diamond storage before creating a PR and after

Perhaps @rndquu might be able to help with this question.

**Comment by PhantomCracker** [comment]
> > Hello @zgorizzo69 , Can you help with something? I have 2 questions:
> > 
> > 1. where can i find the diamond storage?
> > 2. is there a backup for diamond storage?
> > 
> > I am thinking of checking the diamond storage before creating a PR and after
> 
> Perhaps @rndquu might be able to help with this question.

I did it in another way so we are good now, thank you!

**Comment by 0x4007** [comment]
I just realized that a week to create a single `.yml` file seems very steep. 

I will need to lower the time estimate to a day if the current assignee is unable to complete.

**Comment by PhantomCracker** [comment]
I will try to finish as fast as possible.

**Comment by PhantomCracker** [comment]
Need to test a few more things here and I will come back with updated PR

**Comment by Arjuna2022** [comment]
name: Check Diamond Storage
on:
  pull_request:
    branches:
      - main
jobs:
  check_diamond_storage:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Check Diamond Storage Hash
        run: |
          echo "Checking diamond storage hash..."
          before_hash=$(sha256sum diamond_storage)
          git merge --no-ff main
          after_hash=$(sha256sum diamond_storage)
          if [ "$before_hash" != "$after_hash" ]; then
            echo "Diamond storage has been tampered with!"
          else
            echo "Diamond storage has not been tampered with."
          fi
Hello everyone, i was wondering if this code helps i got here from build box. very inexpericeind so any advice  would be kind helpfull.

**Comment by rndquu** [comment]
> name: Check Diamond Storage on: pull_request: branches: - main jobs: check_diamond_storage: runs-on: ubuntu-latest steps: - uses: actions/checkout@v2 - name: Check Diamond Storage Hash run: | echo "Checking diamond storage hash..." before_hash=$(sha256sum diamond_storage) git merge --no-ff main after_hash=$(sha256sum diamond_storage) if [ "$before_hash" != "$after_hash" ]; then echo "Diamond storage has been tampered with!" else echo "Diamond storage has not been tampered with." fi Hello everyone, i was wondering if this code helps i got here from build box. very inexpericeind so any advice would be kind helpfull.

this issue is already taken by the other bounty hunter, pls find an issue with no assignee

**Comment by 0x4007** [comment]
> name: Check Diamond Storage
> on:
> pull_request:
> branches:
> - main
> jobs:
> check_diamond_storage:
> runs-on: ubuntu-latest
> steps:
> - uses: actions/checkout@v2
> - name: Check Diamond Storage Hash
> run: |
> echo "Checking diamond storage hash..."
> before_hash=$(sha256sum diamond_storage)
> git merge --no-ff main
> after_hash=$(sha256sum diamond_storage)
> if [ "$before_hash" != "$after_hash" ]; then
> echo "Diamond storage has been tampered with!"
> else
> echo "Diamond storage has not been tampered with."
> fi
> Hello everyone, i was wondering if this code helps i got here from build box. very inexpericeind so any advice would be kind helpfull.

I think that a fully GitHub Action implementation makes a lot more sense actually vs what we have going on in the associated pull request. I'm just unsure if it makes sense to throw an error on CI if changes happen. I don't think the pull request reviewers should be expected to manually check the CI logs every time before merging.

**Comment by PhantomCracker** [comment]
Hei @pavlovcik,
I can throw an error if changes happen if this is required.
Regarding github action or script, it will be a long action if we write everything there and it will be harder to read.
Like this, we use best practices for this type of issue/feature.

**Comment by 0x4007** [comment]
It might be best to post a comment in the pull request so that we don't need to dig through the CI logs, similar to what we do already for the deployments.

**Comment by rndquu** [comment]
This issue is really tricky

We have a [diamond contract](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/Diamond.sol) which includes all the [facets](https://github.com/ubiquity/ubiquity-dollar/tree/development/packages/contracts/src/dollar/facets).

Each facet has a struct ([example](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibCreditNftManager.sol#L43)) which is used as a storage in a unique slot ([example](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibCreditNftManager.sol#L26)).

So on each PR we should check that:
- state variables in storage structs are not modified
- now new state variables are added in the beginning or middle of storage structs (we can add new state variables only to the end of a storage struct in order not to allow a storage collision)
- check that the same rules are also applied to the [AppStorage](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibAppStorage.sol) (which occupies slot 0 in all of the facets)
- make sure that newly added facets are also checked accordingly

So there should be a script that fetches storage layouts from all of the facets (perhaps via [forge inspect](https://book.getfoundry.sh/reference/forge/forge-inspect)) and compares the structure in the PR with the structure in the `development` branch

**Comment by PhantomCracker** [comment]
@rndquu  I was looking on web3. eth.getStorageAt() but this will not work on us because we do not have the addresses for the libraries;
Also, the forge inspect command always return that empty object (same as on Diamond contract).
Can you provide some help on what do you think that we should use to check libraries storage?
Maybe we should create a custom function that check the storage for each struct?

**Comment by rndquu** [comment]
Regarding the libraries, we should have some function that checks that libs' structs used as a storage ([example](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibCreditNftManager.sol#L43)) do not have newly added properties in the beginning or the middle of the struct in order to eliminate storage collisions on contracts upgrade

> I was looking on web3. eth.getStorageAt() but this will not work on us because we do not have the addresses for the libraries;

Libs with only internal methods are included in the contracts' bytecode hence we still can get storage values via `web3. eth.getStorageAt()` or `forge inspect`: https://docs.soliditylang.org/en/v0.8.15/internals/layout_in_storage.html

> Maybe we should create a custom function that check the storage for each struct?

Perhaps

**Comment by PhantomCracker** [comment]
Hei, so let me clarify the next: for libraries we need to check that the new structs are not modified (new PR branch -> dev branch) and for further development on the actual changes from this PR, create the new function that checks the structs talked before; am I right? :D 
Thank you so much for explanations @rndquu ! 👍

**Comment by rndquu** [comment]
> Hei, so let me clarify the next: for libraries we need to check that the new structs are not modified (new PR branch -> dev branch) and for further development on the actual changes from this PR, create the new function that checks the structs talked before; am I right? :D Thank you so much for explanations @rndquu ! 👍

Structs can be modified but only via addition to the end of them, you should check [this](https://proxies.yacademy.dev/pages/security-guide/#storage-collision-vulnerability) and get to know how it can be mitigated in the diamond proxy pattern

**Comment by 0x4007** [comment]
This unassign seems unexpected. Shouldn't there be follow ups from the bot?

**Comment by PhantomCracker** [comment]
Oh, now i see, thank you pavlo!

**Comment by PhantomCracker** [comment]
Hei @rndquu ,
I want to clarify something here again: Let's say that I have this struct:
struct ChefStorage {  uint256 governancePerBlock;  uint256 governanceMultiplier; uint256 minPriceDiffToUpdateMultiplier; uint256 lastPrice; uint256 governanceDivider;  PoolInfo pool;  mapping(uint256 => StakingShareInfo) ssInfo; uint256 totalShares; }

If I compare this value with the one from development we can accept only something like this?: 
struct ChefStorage {  uint256 governancePerBlock;  uint256 governanceMultiplier; uint256 minPriceDiffToUpdateMultiplier; uint256 lastPrice; uint256 governanceDivider;  PoolInfo pool;  mapping(uint256 => StakingShareInfo) ssInfo; uint256 totalShares; uint256 newVariable; }

Thank you!

**Comment by rndquu** [comment]
> Hei @rndquu , I want to clarify something here again: Let's say that I have this struct: struct ChefStorage { uint256 governancePerBlock; uint256 governanceMultiplier; uint256 minPriceDiffToUpdateMultiplier; uint256 lastPrice; uint256 governanceDivider; PoolInfo pool; mapping(uint256 => StakingShareInfo) ssInfo; uint256 totalShares; }
> 
> If I compare this value with the one from development we can accept only something like this?: struct ChefStorage { uint256 governancePerBlock; uint256 governanceMultiplier; uint256 minPriceDiffToUpdateMultiplier; uint256 lastPrice; uint256 governanceDivider; PoolInfo pool; mapping(uint256 => StakingShareInfo) ssInfo; uint256 totalShares; uint256 newVariable; }
> 
> Thank you!

Code formatting would really help

Yes, new variable is added to the end of the `ChefStorage` struct

**Comment by PhantomCracker** [comment]
Thinking on a method to compare each struct from newPR with each struct from dev; work in progress

**Comment by PhantomCracker** [comment]
This will be finished in about 3-4 days and I will add the PR for review; it takes a little more time than I expected

**Comment by 0x4007** [comment]
> This will be finished in about 3-4 days and I will add the PR for review; it takes a little more time than I expected

How is it looking?

**Comment by PhantomCracker** [comment]
Not the best... still have some issues with checking structs structure, but is work in progress.

**Comment by gitcoindev** [comment]
hi @PhantomCracker I saw your changes, this is a significant effort to implement from scratch indeed.

Perhaps https://github.com/Rubilmax/foundry-storage-check would be worth to check ?

**Comment by PhantomCracker** [comment]
> https://github.com/Rubilmax/foundry-storage-check

Thank you @gitcoindev , I will check it
Later edit: I had a look over it and it looks pretty nice, I will do some tests to see how this works and how we can integrate it in our project.

**Comment by 0x4007** [comment]
> hi @PhantomCracker I saw your changes, this is a significant effort to implement from scratch indeed.
> 
> 
> 
> Perhaps https://github.com/Rubilmax/foundry-storage-check would be worth to check ? 

Thanks for sharing your findings @gitcoindev

**Comment by rndquu** [comment]
> > https://github.com/Rubilmax/foundry-storage-check
> 
> Thank you @gitcoindev , I will check it Later edit: I had a look over it and it looks pretty nice, I will do some tests to see how this works and how we can integrate it in our project.

There is also this [tool](https://docs.openzeppelin.com/upgrades-plugins/1.x/api-core#usage) by openzeppelin which works pretty much the same way as https://github.com/Rubilmax/foundry-storage-check. Although both tools (as far as I understand) don't support diamond proxies perhaps there is a way to somehow modify our contracts dynamically to make them compatible with those tools.

**Comment by PhantomCracker** [comment]
@gitcoindev what do you think about @rndquu last comment?

**Comment by gitcoindev** [comment]
hi @PhantomCracker unfortunately I am not aware of the difference between OpenZeppelin tool and foundry-storage-check, I would start first with in-depth reading of diamond pattern docs https://eips.ethereum.org/EIPS/eip-2535 then compare the two tools.

**Comment by PhantomCracker** [comment]
Back to work in progress then :) got some issues on foundry-storage-check but i am looking into it

**Comment by PhantomCracker** [comment]
@rndquu, @gitcoindev this starts to overwhelm me and I am not sure that I can develop this anymore... It's a little over my skills. Should I release the bounty or maybe can I obtain some more help on this? ...
Thank you and sorry for incoveniences... :(

**Comment by rndquu** [comment]
> @rndquu, @gitcoindev this starts to overwhelm me and I am not sure that I can develop this anymore... It's a little over my skills. Should I release the bounty or maybe can I obtain some more help on this? ... Thank you and sorry for incoveniences... :(

Simply release the bounty back to the devpool and find another one to solve

**Comment by gitcoindev** [comment]
Since the issue was released back to the dev pool I would like to try and have a look , but want to do this properly as this indeed is a complex (but looks doable and fun) task. I give myself a few days for the initial research on diamond storage and available tools and ways to modify them. Then I will start around Thursday / Friday with the implementation after I finalize the approach.

**Comment by gitcoindev** [comment]
Most relevant articles that I have found so far:

https://docs.soliditylang.org/en/v0.8.21/internals/layout_in_storage.html#storage-inplace-encoding
https://dev.to/mudgen/how-diamond-storage-works-90e
https://dev.to/mudgen/solidity-libraries-can-t-have-state-variables-oh-yes-they-can-3ke9
https://medium.com/@0xFluffyBeard/migrating-the-eip-2535-diamond-storage-layout-cb28b9c47e71
https://eip2535diamonds.substack.com/p/understanding-delegatecall-and-how
https://eip2535diamonds.substack.com/p/example-of-a-diamond-upgrade
https://www.certik.com/resources/blog/7laIe0oZGK6IoYDwn0g2Jp-diamond-proxy-contracts-best-practices

**Comment by PhantomCracker** [comment]
If you need help with development, let me know @gitcoindev :)

**Comment by molecula451** [comment]
i think a shell script that gets triggered it's _potentially_ a helper for check things on "changing"

**Comment by molecula451** [comment]
also I think it's best not to **_over-engineer_** things regardless of how complex the tasks look. Over engineering things creates technical debt, instead of adding value, this tasks complex as it seems, it should be implemented in the most simple way

**Comment by molecula451** [comment]
> https://docs.soliditylang.org/en/v0.8.21/internals/layout_in_storage.html#storage-inplace-encoding

I suggest you only rely on Diamond things, as this is an implementation under solidity, structs works different, solidity does not document anything about diamonds.

**Comment by gitcoindev** [comment]
Thanks @molecula451 . I came up with a one - liner using `forge inspect` that outputs storage layouts of all contracts in ubiquity-dollar. Then I had a look at https://github.com/Rubilmax/foundry-storage-check just to see what approach did the author use there and noticed that  this tool is basically a typescript wrapper around `forge inspect` with a GitHub action example against a single contract that stores and compares the previous layout. 

So I will either create a shell script that will run the tool against multiple diamond contracts and shared lib storage or fork it and extend the tool to use multiple contracts, or can also read the one-liner output into Python do the comparison and integrate into GitHub action. I will come up with something over the weekend.

**Comment by molecula451** [comment]
sounds (and seems) like a good approach, use `yarn build` locally many times as you increasingly start putting pieces together

**Comment by molecula451** [comment]
> 

Yes but no Python, the reason it's to no add complexity and more dependencies, just shell and typescript it's acceptable

**Comment by gitcoindev** [comment]
@molecula451 all right, I need to rework then before I open the PR, leave out shell and typescript only.

**Comment by gitcoindev** [comment]
I followed @molecula451 ' s advice and considered @rndquu 's comments.

First I analyzed the complete storage output of all contracts using the following one liner:

`forge tree | grep ^src | cut -d' ' -f1 | xargs basename -s | cut -d'.' -f1 | xargs -I{} forge inspect {} storage | tee -a ubiquity-dollar-storage.txt`

Initially I created a script that did a diff between storage layout files, then wanted do a json comparison of the layouts in Python. Then I read again all comments in the thread.

Quoting @rndquu 

> So on each PR we should check that:

> * state variables in storage structs are not modified
> * now new state variables are added in the beginning or middle of storage structs (we can add new state variables only to the end of a storage struct in order not to allow a storage collision)
> * check that the same rules are also applied to the [AppStorage](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibAppStorage.sol) (which occupies slot 0 in all of the facets)
> * make sure that newly added facets are also checked accordingly
> * So there should be a script that fetches storage layouts from all of the facets (perhaps via [forge inspect](https://book.getfoundry.sh/reference/forge/forge-inspect)) and compares the structure in the PR with the structure in the development branch

The simplified view of the monorepo is that diamond init contract initializes parts of storage of the diamond, then there are multiple facets, each may use its own storage and the storage implementation is located in libraries. Facets can be extended, and added plus there is a shared AppStorage between facets with slot 0. Additionally there is a list of core contracts for which #775 was mentioned.

I did a forge inspect of the individual facets from command line and realized that facets use their libraries storage and when forge inspect layout is executed on a facet it returns the storage inherited from library.

Unfortunately https://github.com/Rubilmax/foundry-storage-check example is about just a single contract with a name hard coded into action yaml but after reading https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs and several articles https://blog.aspect.dev/github-actions-dynamic-matrix , https://tomasvotruba.com/blog/2020/11/16/how-to-make-dynamic-matrix-in-github-actions/ I found a way to cover multiple contracts and was able to create a GitHub action with dynamic matrix that covers those scenarios. The workflow triggers a provide_contracts job which creates a dynamic array of storage checks, including all facets (see example here https://github.com/gitcoindev/ubiquity-dollar/actions/runs/6153043773/workflow ) . The list of contracts is created with a shell script that uses forge and appended into a file. Then I convert the file into and array (actually a JSON object) which is used to trigger multiple GitHub workers in parallel.

How this works:

* whenever there is a push to `development` branch, the GitHub action collects storage layouts using foundry-storage-check and pushes them in the JSON format to job artifacts e.g.:

| Artifact        | Size           |
| ------------- |-------------|
| development.packages_contracts_src_dollar_facets_BondingCurveFacet.sol-BondingCurveFacet.json | 6.29 KB |  
| development.packages_contracts_src_dollar_facets_ChefFacet.sol-ChefFacet.json | 5.95 KB |  
| development.packages_contracts_src_dollar_facets_CollectableDustFacet.sol-CollectableDustFacet.json | 6.42 KB |  
| development.packages_contracts_src_dollar_facets_CreditNftManagerFacet.sol-CreditNftManagerFacet.json | 6.49 KB |  

etc.

see example artifacts on a forked repository:
<br class="Apple-interchange-newline">[development.packages_contracts_src_dollar_facets_BondingCurveFacet.sol-BondingCurveFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580013)
[development.packages_contracts_src_dollar_facets_ChefFacet.sol-ChefFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580014)
[development.packages_contracts_src_dollar_facets_CollectableDustFacet.sol-CollectableDustFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580015)
[development.packages_contracts_src_dollar_facets_CreditNftManagerFacet.sol-CreditNftManagerFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580017)
</br>

This acts as a 'golden' set of storage layouts to compare to against pull requests. 

* Each time a pull request event is triggered, the storage check workflow runs against a feature branch and compares the layout with the one in the development branch and if any violations are found the action automatically adds a comment in the pull request.

To provide a real example I created a pull request on my forked repository here:

https://github.com/gitcoindev/ubiquity-dollar/pull/4/files

As you can see I added a `uint256 findMe` variable in the middle of the storage struct and the storage check workflow issued three warnings about struct variables modified after `findMe` was inserted. 

I will open now a pull request with the solution.

**Comment by molecula451** [comment]
> I followed @molecula451 ' s advice and considered @rndquu 's comments.
> 
> First I analyzed the complete storage output of all contracts using the following one liner:
> 
> `forge tree | grep ^src | cut -d' ' -f1 | xargs basename -s | cut -d'.' -f1 | xargs -I{} forge inspect {} storage | tee -a ubiquity-dollar-storage.txt`
> 
> Initially I created a script that did a diff between storage layout files, then wanted do a json comparison of the layouts in Python. Then I read again all comments in the thread.
> 
> Quoting @rndquu 
> 
> > So on each PR we should check that:
> 
> > * state variables in storage structs are not modified
> > * now new state variables are added in the beginning or middle of storage structs (we can add new state variables only to the end of a storage struct in order not to allow a storage collision)
> > * check that the same rules are also applied to the [AppStorage](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibAppStorage.sol) (which occupies slot 0 in all of the facets)
> > * make sure that newly added facets are also checked accordingly
> > * So there should be a script that fetches storage layouts from all of the facets (perhaps via [forge inspect](https://book.getfoundry.sh/reference/forge/forge-inspect)) and compares the structure in the PR with the structure in the development branch
> 
> The simplified view of the monorepo is that diamond init contract initializes parts of storage of the diamond, then there are multiple facets, each may use its own storage and the storage implementation is located in libraries. Facets can be extended, and added plus there is a shared AppStorage between facets with slot 0. Additionally there is a list of core contracts for which #775 was mentioned.
> 
> I did a forge inspect of the individual facets from command line and realized that facets use their libraries storage and when forge inspect layout is executed on a facet it returns the storage inherited from library.
> 
> Unfortunately https://github.com/Rubilmax/foundry-storage-check example is about just a single contract with a name hard coded into action yaml but after reading https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs and several articles https://blog.aspect.dev/github-actions-dynamic-matrix , https://tomasvotruba.com/blog/2020/11/16/how-to-make-dynamic-matrix-in-github-actions/ I found a way to cover multiple contracts and was able to create a GitHub action with dynamic matrix that covers those scenarios. The workflow triggers a provide_contracts job which creates a dynamic array of storage checks, including all facets (see example here https://github.com/gitcoindev/ubiquity-dollar/actions/runs/6153043773/workflow ) . The list of contracts is created with a shell script that uses forge and appended into a file. Then I convert the file into and array (actually a JSON object) which is used to trigger multiple GitHub workers in parallel.
> 
> How this works:
> 
> * whenever there is a push to `development` branch, the GitHub action collects storage layouts using foundry-storage-check and pushes them in the JSON format to job artifacts e.g.:
> 
> | Artifact        | Size           |
> | ------------- |-------------|
> | development.packages_contracts_src_dollar_facets_BondingCurveFacet.sol-BondingCurveFacet.json | 6.29 KB |  
> | development.packages_contracts_src_dollar_facets_ChefFacet.sol-ChefFacet.json | 5.95 KB |  
> | development.packages_contracts_src_dollar_facets_CollectableDustFacet.sol-CollectableDustFacet.json | 6.42 KB |  
> | development.packages_contracts_src_dollar_facets_CreditNftManagerFacet.sol-CreditNftManagerFacet.json | 6.49 KB |  
> 
> etc.
> 
> see example artifacts on a forked repository:
> <br class="Apple-interchange-newline">[development.packages_contracts_src_dollar_facets_BondingCurveFacet.sol-BondingCurveFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580013)
> [development.packages_contracts_src_dollar_facets_ChefFacet.sol-ChefFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580014)
> [development.packages_contracts_src_dollar_facets_CollectableDustFacet.sol-CollectableDustFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580015)
> [development.packages_contracts_src_dollar_facets_CreditNftManagerFacet.sol-CreditNftManagerFacet.json](https://github.com/gitcoindev/ubiquity-dollar/suites/16057941853/artifacts/916580017)

I am unable to visualize this .json files to gather the whole context. It shows 404 can You recheck

**Comment by molecula451** [comment]
> As you can see I added a `uint256 findMe` variable in the middle of the storage struct and the storage check workflow issued three warnings about struct variables modified after `findMe` was inserted. 

I am checking your forked repo but i am not seeing the triggered Workflow besides the build, the slither, and coverage failing and not within this context

**Comment by molecula451** [comment]
> Each time a pull request event is triggered, the storage check workflow runs against a feature branch and compares the layout with the one in the development branch and if any violations are found the action automatically adds a comment in the pull request.

This sounds nice Mimick this in your forked repo so we all may see

**Comment by gitcoindev** [comment]
Hi @molecula451 thank you for starting the review! I invited via GitHub you, @rndquu , @pavlovcik and @PhantomCracker to collaborators of my fork, hopefully you will be able to access all links after you accept the invite. In case you still get 404 after you accept the invite please let me know and I will try to figure out how to fix this.

**Comment by molecula451** [comment]
> Hi @molecula451 thank you for starting the review! I invited via GitHub you, @rndquu , @pavlovcik and @PhantomCracker to collaborators of my fork, hopefully you will be able to access all links after you accept the invite. In case you still get 404 after you accept the invite please let me know and I will try to figure out how to fix this.

I'm seeing the storage layout Workflow but make it find a defect? I am not seeing the coment on the PR that something it's wrong

**Comment by gitcoindev** [comment]
> > Hi @molecula451 thank you for starting the review! I invited via GitHub you, @rndquu , @pavlovcik and @PhantomCracker to collaborators of my fork, hopefully you will be able to access all links after you accept the invite. In case you still get 404 after you accept the invite please let me know and I will try to figure out how to fix this.
> 
> I'm seeing the storage layout Workflow but make it find a defect? I am not seeing the coment on the PR that something it's wrong

You should be able to see warnings in the changed files tab, let me paste the screenshot:

![Screenshot from 2023-09-12 21-27-51](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/8d32fbb5-3e48-4e85-ae76-6c5022c2fa71)

**Comment by molecula451** [comment]
> > > Hi @molecula451 thank you for starting the review! I invited via GitHub you, @rndquu , @pavlovcik and @PhantomCracker to collaborators of my fork, hopefully you will be able to access all links after you accept the invite. In case you still get 404 after you accept the invite please let me know and I will try to figure out how to fix this.
> > 
> > I'm seeing the storage layout Workflow but make it find a defect? I am not seeing the coment on the PR that something it's wrong
> 
> You should be able to see warnings in the changed files tab, let me paste the screenshot:
> 
> ![Screenshot from 2023-09-12 21-27-51](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/8d32fbb5-3e48-4e85-ae76-6c5022c2fa71)
> 

@rndquu this has started to looking good on that screenshot

@gitcoindev now can you trigger it the actual PR You may push a bad commit then we all Review then we push a good commit when finish only if need it

**Comment by molecula451** [comment]
Also make it trigger on one of the facets

**Comment by gitcoindev** [comment]
Triggered on a commit that moves to end of the struct caused the warnings to disappear

![Screenshot from 2023-09-12 22-04-05](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/fa784753-e97b-487e-95c5-c26a74be3b5b)

**Comment by molecula451** [comment]
> Triggered on a commit that moves to end of the struct caused the warnings to disappear
> 
> ![Screenshot from 2023-09-12 22-04-05](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/fa784753-e97b-487e-95c5-c26a74be3b5b)
> 
> 
> 

But trigger it on #788

**Comment by gitcoindev** [comment]
Third attempt against facet change - no comment was added ,  good catch! I need to debug this tomorrow and compare solc storage layout output as well, the magic in foundry happens here https://github.com/foundry-rs/foundry/blob/master/crates/cast/bin/cmd/storage.rs#L242 this may require a few hours of debugging.

**Comment by gitcoindev** [comment]
> > Triggered on a commit that moves to end of the struct caused the warnings to disappear
> > ![Screenshot from 2023-09-12 22-04-05](https://user-images.githubusercontent.com/88761781/267451934-fa784753-e97b-487e-95c5-c26a74be3b5b.png)
> 
> But trigger it on #788

#788 will start detecting correctly only after merging, this is completely normal and described in foundry-storage-check

ℹ️ An error will appear at first run!
🔴 Error: No workflow run found with an artifact named "..."
As the action is expecting a comparative file stored on the base branch and cannot find it (because the action never ran on the target branch and thus has never uploaded any storage layout)

**Comment by molecula451** [comment]
> Triggered on a commit that moves to end of the struct caused the warnings to disappear

But thats because it takes it as if You adding a new var and not touching the actual ones?

**Comment by gitcoindev** [comment]
Ok so to sum up - I need to investigate tomorrow why facet comment was not added correctly.

**Comment by gitcoindev** [comment]
> > Triggered on a commit that moves to end of the struct caused the warnings to disappear
> 
> But thats because it takes it as if You adding a new var and not touching the actual ones?

Yes exactly, I showed this to prove that the checker works when the issue is solved in a new commit. Adding a new var at the end of the struct is perfectly fine and should not trigger a warning.

**Comment by molecula451** [comment]
Good. Let us know

**Comment by molecula451** [comment]
> Third attempt against facet change - no comment was added ,  good catch! I need to debug this tomorrow and compare solc storage layout output as well, the magic in foundry happens here https://github.com/foundry-rs/foundry/blob/master/crates/cast/bin/cmd/storage.rs#L242 this may require a few hours of debugging.

👆🏻,

**Comment by gitcoindev** [comment]
> Good. Let us know

I had to deep dive to understand the findings. For Diamond type of contracts Solidity compiler (solc) returns empty storage table when asked, because it does not recognize correctly the storage pointers returned by Libraries, which are a special case and 'Library can't have state variables', which we know can, and is described at https://dev.to/mudgen/solidity-libraries-can-t-have-state-variables-oh-yes-they-can-3ke9 .

Moreover I found out that there are efforts for the tool chain to recognize correctly Diamond storage using well defined namespaces : EIP-7021 (draft) for standardizing storage namespaces https://eips.ethereum.org/EIPS/eip-7201
discussed here as well https://ethereum-magicians.org/t/eip-7201-namespaced-storage-layout/14796/10

So currently there is no simple way to get Diamond storage layout using build tool chain.

But I believe that the current issue can be solved *somehow* so I dug deeper.

1. foundry-check-storage tool uses `foundry inspect storage-layout` to inspect storage layout.
2. Foundry uses Solidity compiler (solc) to get storage layout.
3. Solidity compiler (solc) implementation to generate the layout is here: https://github.com/ethereum/solidity/blob/64a0f6270014a8d275b657e53e4cde72e3d76a3d/libsolidity/interface/StorageLayout.cpp#L47

One way to try to solve the issue would be to patch Solidity compiler, to generate also storage struct from the libraries but I am against this approach, because:

1. It is not standardized, the proper solution should be provided after EIP-7021 is formalized.
2. Would require patch maintenance with each solc upgrade.
3. Sounds hackish as hell.

Looking at the solc sources I realized the Solidity compiler must know everything about the internal structs and this information is available in the AST (Abstract Syntax Tree) representation of the source code.

I changed locally one of the libraries that return storage to the facet and was able to generate AST using:

`solc @openzeppelin=lib/openzeppelin-contracts abdk/=lib/abdk-libraries-solidity/ src/dollar/libraries/LibCollectableDust.sol --pretty-json --ast-compact-json > lol.txt`

in which I found

```
{
          "canonicalName": "LibCollectableDust.Tokens",
          "id": 52,
          "members":
          [
            {
              "constant": false,
              "id": 49,
              "mutability": "mutable",
              "name": "protocolTokens",
              "nameLocation": "1153:14:15",
              "nodeType": "VariableDeclaration",
              "scope": 52,
              "src": "1128:39:15",
              "stateVariable": false,  <-- BINGO
              "storageLocation": "default",
              "typeDescriptions":
              {
                "typeIdentifier": "t_struct$_AddressSet_$1022_storage_ptr",
                "typeString": "struct EnumerableSet.AddressSet" <-- first field of the struct
              },
              "typeName":
              {
                "id": 48,
                "nodeType": "UserDefinedTypeName",
                "pathNode":
                {
                  "id": 47,
                  "name": "EnumerableSet.AddressSet",
                  "nameLocations":
                  [
                    "1128:13:15",
                    "1142:10:15"
                  ],
                  "nodeType": "IdentifierPath",
                  "referencedDeclaration": 1022,
                  "src": "1128:24:15"
                },
                "referencedDeclaration": 1022,
                "src": "1128:24:15",
                "typeDescriptions":
                {
                  "typeIdentifier": "t_struct$_AddressSet_$1022_storage_ptr",
                  "typeString": "struct EnumerableSet.AddressSet"
                }
              },
              "visibility": "internal"
            },
            {
              "constant": false,
              "id": 51,
              "mutability": "mutable",
              "name": "wtfTokens", <-- second field of the struct that I added locally
              "nameLocation": "1185:9:15",
              "nodeType": "VariableDeclaration",
              "scope": 52,
              "src": "1177:17:15",
              "stateVariable": false,
              "storageLocation": "default",
              "typeDescriptions":
              {
                "typeIdentifier": "t_uint256",
                "typeString": "uint256"
              },
              "typeName":
              {
                "id": 50,
                "name": "uint256",
                "nodeType": "ElementaryTypeName",
                "src": "1177:7:15",
                "typeDescriptions":
                {
                  "typeIdentifier": "t_uint256",
                  "typeString": "uint256"
                }
              },
              "visibility": "internal"
            }
          ],
          "name": "Tokens",
          "nameLocation": "1111:6:15",
          "nodeType": "StructDefinition",
          "scope": 203,
          "src": "1104:97:15",
          "visibility": "public"
        },
        {
          "constant": true,
          "documentation":
          {
            "id": 53,
            "nodeType": "StructuredDocumentation",
            "src": "1207:60:15",
            "text": "@notice Storage slot used to store data for this library"
          },
```

I still need to think if and how to extract this information, because this can be a basis to detect changes in the Diamond structs directly from AST and later add comments in the CI.

**Comment by gitcoindev** [comment]
This will be complex but not impossible. I keep analyzing AST. I took a more complex example, LibDiamond.sol.
This library contains three different structs and the convention is that for Diamond Libraries only one struct is returned as a storage. The trick is to find which one.

I discovered that for Diamond Storage libraries one thing stands out. There is always only one function, that in its returnParameters AST contains `"storageLocation": "storage"` entry. If this entry is available, `"typeDescriptions"` contains a pointer to the actual storage, and `typeName["name"]` field contains name of the struct used as a storage.

```
"returnParameters":
          {
            "id": 67,
            "nodeType": "ParameterList",
            "parameters":
            [
              {
                "constant": false,
                "id": 66,
                "mutability": "mutable",
                "name": "ds",
                "nameLocation": "2346:2:5",
                "nodeType": "VariableDeclaration",
                "scope": 74,
                "src": "2323:25:5",
                "stateVariable": false,
                "storageLocation": "storage",
                "typeDescriptions":
                {
                  "typeIdentifier": "t_struct$_DiamondStorage_$61_storage_ptr",
                  "typeString": "struct LibDiamond.DiamondStorage"
                },
                "typeName":
                {
                  "id": 65,
                  "nodeType": "UserDefinedTypeName",
                  "pathNode":
                  {
                    "id": 64,
                    "name": "DiamondStorage",
                    "nameLocations":
                    [
                      "2323:14:5"
                    ],
                    "nodeType": "IdentifierPath",
                    "referencedDeclaration": 61,
                    "src": "2323:14:5"
                  },
                  "referencedDeclaration": 61,
                  "src": "2323:14:5",
                  "typeDescriptions":
                  {
                    "typeIdentifier": "t_struct$_DiamondStorage_$61_storage_ptr",
                    "typeString": "struct LibDiamond.DiamondStorage"
                  }
                },
                "visibility": "internal"
              }
            ],
            "src": "2322:27:5"
          },
```

**Comment by gitcoindev** [comment]
So, in order to make the solution complete will be to:

1. Find all Solidity libraries in `contracts/src/dollar/libraries`
2. Find which libraries are Diamond libraries, i.e. they have to contain a function with return parameter `"storageLocation": "storage"` and from this function extract name of the returned struct. This struct is a Diamond storage.
3. Create from AST the storage layout from the struct found in step 2.
4. The storage layout from step 3 must be added to the same list that is already analysed in the GitHub action.

**Comment by molecula451** [comment]
it's possible that you are heading the complex path, maybe another simple way round?

**Comment by gitcoindev** [comment]
@molecula451 you might be right -) I am giving solidity-parser-antlr a try. If I do not manage to get my original approach done by Sunday I will try something less advanced. I considered several options and unfortunately using parsed AST will be a must anyway. Solidity compiler returns empty storage layout for diamond libraries and until that changes we cannot compare its storage layout output to anything. So fingers crossed and I am back to work!

**Comment by molecula451** [comment]
simple its better so others can maintain the implementation

**Comment by gitcoindev** [comment]
Just an update: working on a fork , trying to simplify my implementation. I applied TDD - add tests first to make sure I did not break anything.

My current state is at https://github.com/gitcoindev/foundry-storage-check/commit/88e425c82f50c47c5f867d66cd0e7ff00d76e1f2 and I still hope to finish by Sunday EOD.

**Comment by gitcoindev** [comment]
Good news! I was able to finish, the Diamond storage checker seems to work correctly now.
I changed the pull request status to 'ready for review'.

I did three releases of the checker on the way, implemented 6 test cases and verified the output on my fork as well. At first I started with elementary structs, then added user defined types supports and in the end mapping support.

My fork changes can be viewed here:

https://github.com/Rubilmax/foundry-storage-check/compare/main...gitcoindev:foundry-storage-check:main

New test cases related to Diamond storage pass locally and on CI:

```
$ yarn test tests/check.test.ts -t "Diamond"
yarn run v1.22.19
$ jest --coverage tests/check.test.ts -t Diamond
 PASS  tests/check.test.ts

...
    Diamond storage
      ✓ should not raise diff when changing name of struct (16 ms)
      ✓ should not raise diff when extending struct (10 ms)
      ✓ should raise type diff when changing type of variable (10 ms)
      ✓ should raise label diff when switching struct variables (10 ms)
      ✓ should raise label diff when inserting custom type struct variables (14 ms)
      ✓ should raise label diff when inserting mapping into struct variables (12 ms)

...

-----------|---------|----------|---------|---------|-----------------------------------------------------------------
File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                               
-----------|---------|----------|---------|---------|-----------------------------------------------------------------
All files  |   73.43 |    56.25 |   92.59 |   73.74 |                                                                 
 check.ts  |    60.9 |    46.55 |   83.33 |   61.16 | 37-63,88-89,165-170,175-180,215-224,226-234,239,255-266,288-313 
 format.ts |   72.22 |    53.33 |     100 |   70.58 | 62-74,86-92                                                     
 input.ts  |   94.64 |    80.95 |     100 |   96.07 | 131-133                                                         
 types.ts  |     100 |      100 |     100 |     100 |                                                                 
-----------|---------|----------|---------|---------|-----------------------------------------------------------------
Test Suites: 1 passed, 1 total
Tests:       19 skipped, 6 passed, 25 total
Snapshots:   0 total
Time:        2.631 s, estimated 6 s
Ran all test suites matching /tests\/check.test.ts/i with tests matching "Diamond".
Done in 3.24s.
```

total run: all tests passing, no regressions:

```
-----------|---------|----------|---------|---------|---------------------
File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s   
-----------|---------|----------|---------|---------|---------------------
All files  |   86.97 |    81.25 |   96.29 |   88.26 |                     
 check.ts  |   83.63 |    84.48 |   91.66 |   85.43 | 175-180,239,288-313 
 format.ts |   77.77 |    66.66 |     100 |   76.47 | 68-74,86-92         
 input.ts  |   94.64 |    80.95 |     100 |   96.07 | 131-133             
 types.ts  |     100 |      100 |     100 |     100 |                     
-----------|---------|----------|---------|---------|---------------------
Test Suites: 2 passed, 2 total
Tests:       27 passed, 27 total
Snapshots:   0 total
Time:        5.811 s
Ran all test suites.
Done in 6.36s.
```

I updated the GitHub action yaml to use forked checker for Diamond storage and changed GitHub action step name as requested during the pull request review.

I will also paste a few screenshots from GitHub action to prove that the checker works.

I will also implement #775 and open a pr today as this issue is related.

**Comment by gitcoindev** [comment]
Example 1: 

![image](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/efad60cb-a018-4baa-9353-c767c60b09ba)

Warnings fixed with commit https://github.com/gitcoindev/ubiquity-dollar/pull/4/commits/d032427aa0a4b9ad8d2a72fd6f1a4d2d7984492e

Example 2:

![image](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/8c4f707f-3045-482d-82ea-c62842679b82)

Warnings fixed with commit https://github.com/gitcoindev/ubiquity-dollar/pull/4/commits/4c2d655605709afc722fc3278a076a1b29af2f44

Final example, all jobs green after fixes

![image](https://github.com/ubiquity/ubiquity-dollar/assets/88761781/99bf1585-6b06-4e93-854d-6263979660d6)

**Comment by gitcoindev** [comment]
Adding a comment for the bot: waiting for PR approval and merge.

**Comment by gitcoindev** [comment]
Hi @rndquu thank you for self-assigning the review. I hope we can have both PRs review and merged soon as I would like to pick up a new task and I am currently blocked.

**Comment by molecula451** [comment]
it's possible that this task needs more iteration as the last closed PR, required a lot of iteration to evaluate a measure the in and outs of the task, you should be able to pick another task it it gets approval

**Comment by rndquu** [comment]
> Hi @rndquu thank you for self-assigning the review. I hope we can have both PRs review and merged soon as I would like to pick up a new task and I am currently blocked.

Hey, sorry for the delay, I'll check everything today

Meanwhile you're free to self assign any issue, as far as I remember if there are pending reviews then the bot will allow to self assign. Anyway you can simply select an issue you want to solve and someone from the core team will assign you manually.

**Comment by molecula451** [comment]
@gitcoindev you may comment on any other task you like to work and you'll be assigned

**Comment by gitcoindev** [comment]
Thanks @rndquu @molecula451 @pavlovcik this was a significant effort not only for me but also for you, checking, testing and commenting took a lot of time. 

Now it seems that the bot was overwhelmed by all of this, seeing this number of comments and did not trigger any comment nor triggered any incentives . I hope we can get this fixed and make the bot generate all incentives correctly.

**Comment by 0x4007** [comment]
We are texting in the group chat about it now t.me/ubiquitydao

Please know that this is a priority for us to address this 

We are scrambling for a fix now 

Thanks for your contributions and your patience.

**Comment by gitcoindev** [comment]
Thanks @rndquu ! I guess the same approach can be used with #775

**Comment by molecula451** [comment]
also looks like it missed lot of the conversational rewars, looks like couldn't handle many commenting?. Would be nice see it re-triggered @rndquu

**Comment by PhantomCracker** [comment]
Congratulations @gitcoindev for finishing this issue! Pretty hard one :)

**Comment by rndquu** [comment]
> also looks like it missed lot of the conversational rewars, looks like couldn't handle many commenting?. Would be nice see it re-triggered @rndquu

We will re-trigger when the bot is switched back to the `main` prod branch. Right now the bot's prod deployment is on a commit without proper conversation rewards.

**Comment by molecula451** [comment]
@rndquu i think we should try to re-trigger conversational rewards here, i think the bot is reacting to those on your calls, there were helpful comment here to reach the goal (which after a couple of days) the PR was nice

**Comment by rndquu** [comment]
> @rndquu i think we should try to re-trigger conversational rewards here, i think the bot is reacting to those on your calls, there were helpful comment here to reach the goal (which after a couple of days) the PR was nice

I don't really want to re-trigger anything because rewards generation is kind of unpredicted right now. Let's wait for a stable bot's release.
